src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp

```diff
@@ -138,66 +187,147 @@ Json::Value CascadiaSettings::ToJson() const
     }
 
-    // TODO:MSFT:20737698 - Display an error if we failed to parse settings
-    // What should we do here if these keys aren't found?For default profile,
-    //      we could always pick the first  profile and just set that as the default.
-    // Finding no schemes is probably fine, unless of course one profile
-    //      references a scheme.  We could fail with come error saying the
-    //      profiles file is corrupted.
-    // Not having any profiles is also bad - should we say the file is corrupted?
-    //      Or should we just recreate the default profiles?
-
-    auto& resultSchemes = resultPtr->_globals.GetColorSchemes();
     if (auto schemes{ json[SchemesKey.data()] })
     {
         for (auto schemeJson : schemes)
         {
             if (schemeJson.isObject())
             {
-                auto scheme = ColorScheme::FromJson(schemeJson);
-                resultSchemes.emplace_back(std::move(scheme));
+                _LayerOrCreateColorScheme(schemeJson);
             }
         }
     }
 
-    if (auto profiles{ json[ProfilesKey.data()] })
+    for (auto profileJson : _GetProfiles(json))
     {
-        for (auto profileJson : profiles)
+        if (profileJson.isObject())
         {
-            if (profileJson.isObject())
-            {
-                auto profile = Profile::FromJson(profileJson);
-                resultPtr->_profiles.emplace_back(profile);
-            }
+            _LayerOrCreateProfile(profileJson);
         }
     }
+}
 
-    return resultPtr;
+// Method Description:
+// - Given a partial json serialization of a Profile object, either layers that
+//   json on a matching Profile we already have, or creates a new Profile
+//   object from those settings.
+// Arguments:
+// - json: an object which should be a partial serialization of a Profile object.
+// Return Value:
+// - <none>
+void CascadiaSettings::_LayerOrCreateProfile(const Json::Value& profileJson)
+{
+    // Layer the json on top of an existing profile, if we have one:
+    auto pProfile = _FindMatchingProfile(profileJson);
+    if (pProfile)
+    {
+        pProfile->LayerJson(profileJson);
+    }
+    else
+    {
+        auto profile = Profile::FromJson(profileJson);
+        _profiles.emplace_back(profile);
+    }
+}
+
+// Method Description:
+// - Finds a profile from our list of profiles that matches the given json
+//   object. Uses Profile::ShouldBeLayered to determine if the Json::Value is a
+//   match or not. This method should be used to find a profile to layer the
+//   given settings upon.
+// - Returns nullptr if no such match exists.
+// Arguments:
+// - <none>
+// Return Value:
+// - a Profile that can be layered with the given json object, iff such a
+//   profile exists.
+Profile* CascadiaSettings::_FindMatchingProfile(const Json::Value& profileJson)
+{
+    for (auto& profile : _profiles)
+    {
+        if (profile.ShouldBeLayered(profileJson))
+        {
+            return &profile;

```

Comment:
I almost feel like it should be `std::optional<std::shared_ptr<Profile>>`

Reply:
This _is_ scary and definitely bit me when I first wrote it.

Would adding `enable_shared_from_this` to Profile (and ColorScheme) below be _less_ scary though? Right now since we only have a `vector<Profile>` it's a little trickier to change that to a `vector<shared_ptr<Profile>>`. Doable, but might push this pr past Friday TBH. It'll have effects that spill into TerminalPage

Reply:
Meh, we can live with it.